Add more `DAGCircuit::copy_empty_like`-related methods

jakelishman opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

The main aim of this PR is to add `DAGCircuit::physical_empty_like_with_capacity`, which is used by anything in Rust space that needs to "rebuild" a DAG to be physical (e.g. Sabre, in #14317).  The first commit is just adding all the methods on the way to that goal.

The "it's error prone" in the second commit message refers to the bug fixed by #14488.

### Details and comments

#### Add capacity-setting `DAGCircuit::copy_empty_like` constructors

`petgraph` currently does not expose any APIs to expand the capacity of the `Vec`s backing `Graph` after construction.  This motivates allowing `copy_empty_like` to accept capacities.

#### Add physical-circuit variant of `DAGCircuit::copy_empty_like`

The new `DAGCircuit::physical_empty_like_with_capacity` is similar in spirit to `copy_empty_like_with_capacity`, except that the output is defined over a certain number of physical qubits.  This operation is common (and error-prone) in layout passes, so is centralised here.

jakelishman committed:
Add capacity-setting `DAGCircuit::copy_empty_like` constructors

`petgraph` currently does not expose any APIs to expand the capacity of
the `Vec`s backing `Graph` after construction.  This motivates allowing
`copy_empty_like` to accept capacities.

jakelishman committed:
Add physical-circuit variant of `DAGCircuit::copy_empty_like`

The new `DAGCircuit::physical_empty_like_with_capacity` is similar in
spirit to `copy_empty_like_with_capacity`, except that the output is
defined over a certain number of physical qubits.  This operation is
common (and error-prone) in layout passes, so is centralised here.

jakelishman committed:
Merge branch 'main' into dag-copy-empty-like

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -4504,6 +4436,121 @@ impl DAGCircuit {
+        )?;
+        out.remove_qubits((0..out.num_qubits()).map(Qubit::new))?;
+        out.qargs_interner = Interner::new();
+        out.add_qreg(QuantumRegister::new_owning("q", num_qubits as u32))?;
```
:see_no_evil:

jakelishman replied:
For better or worse, this _does_ define a physical circuit.

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -4504,6 +4436,121 @@ impl DAGCircuit {
+            num_edges + extra_qubits,
+            "alike",
+        )?;
+        out.remove_qubits((0..out.num_qubits()).map(Qubit::new))?;
```
This does seem less than ideal to create a new dag delete all the bits and then add the new register. I doubt it will be a real bottleneck for anything except for really wide circuits. But I'm surprised you didn't just add a flag to skip adding the bit operations or separate method that excludes adding them. I'd probably prefer to not have to debug the performance issue in the future and just do the more efficient thing now.

jakelishman replied:
I don't actually think the performance impact would be measurable - the circuit is necessarily empty, so the "remove bits" operation is cheap.  Even with the slight awkwardness here, the method is still entirely linear in the circuit width, and making this step a little more efficient doesn't change that.

That said, yeah, I can work out some other API to do it.  Mostly, the trick is that `copy_empty_like` has to do quite a bunch of work that's quite error-prone to get right, so I don't want to duplicate any of its logic; I want that all entirely centralised.

jakelishman replied:
Done in 9178fe6ffa.

mtreinish commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -4504,6 +4436,121 @@ impl DAGCircuit {
+    ///
+    /// This method is intended for use by passes that are converting a virtual DAG to a physical
+    /// one.
+    pub fn physical_empty_like_with_capacity(
```
Do you want to have a unit test for this just so we exercise it in tests.

jakelishman replied:
This isn't directly exposed to Python, so it's a bit of nuisance to do it well.  I can do it in Rust if you like for now. The majority of the logic is tested via the Python-space unit tests of `copy_empty_like`, and the rest of `physical_empty_like` will be done by integration with Sabre in the follow-up PRs, if you don't mind so strongly.

mtreinish replied:
I actually did mean a rust unit test here, sorry that wasn't clear. It's not critical either way because I agree it will be tested implicitly via Python in the followup PR. I was thinking it'd be good to get in the habit of more direct rust space testing in cases where we're adding functionality not exercised anywhere else.

jakelishman replied:
Done in 6d6508cb1c.

mtreinish reviewed (commented):
Overall this LGTM. But I think we should just fix the performance issue in `physical_empty_like_with_capacity` now just so we don't have to worry about it. If this was embedded in the sabre PR I'd probably have let it slide, but as a standalone PR it makes more sense to just do it correctly now.

jakelishman committed:
Add `DAGCircuit::qubitless_empty_like_with_capacity`

This is the base method of all the other `*_empty_like_*` methods.  We
don't need the full generality of allowing other classical data to be
present or not; we don't know of any uses of that within the library.

jakelishman committed:
Restore interner-copying behaviour

jakelishman committed:
Include comment on Interner copying

jakelishman committed:
Merge remote-tracking branch 'ibm/main' into dag-copy-empty-like

jakelishman committed:
Make `DAGCircuit::name` and `metadata` public

These are logically entirely public from Python space and have no
data-coherence troubles.  This allows Rust to set them as well.

jakelishman committed:
Add Rust-space test of `physical_empty_like`

mtreinish reviewed (approved):
LGTM, thanks for the updates
